File: src/main/java/com/refinedmods/refinedstorage/network/grid/PortableGridFluidDeltaMessage.java
package com.refinedmods.refinedstorage.network.grid;

import com.refinedmods.refinedstorage.api.util.StackListResult;
import com.refinedmods.refinedstorage.screen.BaseScreen;
import com.refinedmods.refinedstorage.screen.grid.GridScreen;
import com.refinedmods.refinedstorage.screen.grid.stack.IGridStack;
import com.refinedmods.refinedstorage.blockentity.grid.portable.IPortableGrid;
import com.refinedmods.refinedstorage.util.StackUtils;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.network.NetworkEvent;
import org.apache.commons.lang3.tuple.Pair;

import java.util.LinkedList;
import java.util.List;
import java.util.function.Supplier;

public class PortableGridFluidDeltaMessage {
    private IPortableGrid portableGrid;
    private List<StackListResult<FluidStack>> deltas;

    private List<Pair<IGridStack, Integer>> clientDeltas;

    public PortableGridFluidDeltaMessage(IPortableGrid portableGrid, List<StackListResult<FluidStack>> deltas) {
        this.portableGrid = portableGrid;
        this.deltas = deltas;
    }

    public PortableGridFluidDeltaMessage(List<Pair<IGridStack, Integer>> clientDeltas) {
        this.clientDeltas = clientDeltas;
    }

    public static PortableGridFluidDeltaMessage decode(FriendlyByteBuf buf) {
        int size = buf.readInt();

        List<Pair<IGridStack, Integer>> clientDeltas = new LinkedList<>();

        for (int i = 0; i < size; ++i) {
            int delta = buf.readInt();

            clientDeltas.add(Pair.of(StackUtils.readFluidGridStack(buf), delta));
        }

        return new PortableGridFluidDeltaMessage(clientDeltas);
    }

    public static void encode(PortableGridFluidDeltaMessage message, FriendlyByteBuf buf) {
        buf.writeInt(message.deltas.size());

        for (StackListResult<FluidStack> delta : message.deltas) {
            buf.writeInt(delta.getChange());

            StackUtils.writeFluidGridStack(buf, delta.getStack(), delta.getId(), null, false, message.portableGrid.getFluidStorageTracker().get(delta.getStack()));
        }
    }

    public static void handle(PortableGridFluidDeltaMessage message, Supplier<NetworkEvent.Context> ctx) {
        BaseScreen.executeLater(GridScreen.class, grid -> {
            message.clientDeltas.forEach(p -> grid.getView().postChange(p.getLeft(), p.getRight()));
        });

        ctx.get().setPacketHandled(true);
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/network/grid/PortableGridFluidDeltaMessage.kt
package com.refinedmods.refinedstorage.network.grid

import com.refinedmods.refinedstorage.api.util.StackListResult
import com.refinedmods.refinedstorage.screen.BaseScreen
import com.refinedmods.refinedstorage.screen.grid.GridScreen
import com.refinedmods.refinedstorage.screen.grid.stack.IGridStack
import com.refinedmods.refinedstorage.tile.grid.portable.IPortableGrid
import com.refinedmods.refinedstorage.util.StackUtils.readFluidGridStack
import com.refinedmods.refinedstorage.util.StackUtils.writeFluidGridStack
import net.minecraft.network.PacketByteBuf
import net.minecraftforge.fluids.FluidInstance
import net.minecraftforge.fml.network.NetworkEvent
import org.apache.commons.lang3.tuple.Pair
import java.util.*
import java.util.function.Consumer
import java.util.function.Supplier

class PortableGridFluidDeltaMessage {
    private var portableGrid: IPortableGrid? = null
    private var deltas: List<StackListResult<FluidInstance>>? = null
    private var clientDeltas: List<Pair<IGridStack, Int>>? = null

    constructor(portableGrid: IPortableGrid?, deltas: List<StackListResult<FluidInstance>>?) {
        this.portableGrid = portableGrid
        this.deltas = deltas
    }

    constructor(clientDeltas: List<Pair<IGridStack, Int>>?) {
        this.clientDeltas = clientDeltas
    }

    companion object {
        fun decode(buf: PacketByteBuf): PortableGridFluidDeltaMessage {
            val size: Int = buf.readInt()
            val clientDeltas: MutableList<Pair<IGridStack, Int>> = LinkedList()
            for (i in 0 until size) {
                val delta: Int = buf.readInt()
                clientDeltas.add(Pair.of(readFluidGridStack(buf), delta))
            }
            return PortableGridFluidDeltaMessage(clientDeltas)
        }

        fun encode(message: PortableGridFluidDeltaMessage, buf: PacketByteBuf) {
            buf.writeInt(message.deltas!!.size)
            for (delta in message.deltas!!) {
                buf.writeInt(delta.change)
                writeFluidGridStack(buf, delta.stack, delta.id, null, false, message.portableGrid!!.fluidStorageTracker[delta.stack])
            }
        }

        fun handle(message: PortableGridFluidDeltaMessage, ctx: Supplier<NetworkEvent.Context>) {
            BaseScreen.executeLater(GridScreen::class.java) { grid: GridScreen ->
                message.clientDeltas!!.forEach(Consumer { p: Pair<IGridStack, Int> -> grid.view.postChange(p.left, p.right) })
                grid.view.sort()
            }
            ctx.get().setPacketHandled(true)
        }
    }
}